src/dfx/src/lib/deps.rs

```diff
@@ -0,0 +1,276 @@
+        self.canisters.insert(canister_id, InitItem::default());
+    }
+
+    pub fn contains(&self, canister_id: &Principal) -> bool {
+        self.canisters.contains_key(canister_id)
+    }
+
+    pub fn get_arg_raw(&self, canister_id: &Principal) -> DfxResult<Vec<u8>> {
+        let init_item = self.canisters.get(canister_id).ok_or_else(|| {
+            anyhow!(
+                "Failed to find {0} entry in init.json. Please run `dfx deps init {0}`.",
+                canister_id
+            )
+        })?;
+        match &init_item.arg_raw {
+            Some(s) => Ok(hex::decode(s)?),
+            None => Ok(vec![]),
+        }
+    }
+}
+
+#[context("Failed to get pull canisters defined in dfx.json.")]
+pub fn get_pull_canisters_in_config(
+    env: &dyn Environment,
+) -> DfxResult<BTreeMap<String, Principal>> {
+    Ok(env
+        .get_config_or_anyhow()?
+        .get_config()
+        .get_pull_canisters()?)
+}
+
+// 1. whether pulled.json is consistent with dfx.json
+// 2. whether downloaded wasm modules are consistent with pulled.json
+pub fn validate_pulled(
+    pulled_json: &PulledJson,
+    pull_canisters_in_config: &BTreeMap<String, Principal>,
+) -> DfxResult {
+    for (name, canister_id) in pull_canisters_in_config {
+        let pulled_canister = pulled_json
+            .canisters
+            .get(canister_id)
+            .ok_or_else(|| anyhow!("Failed to find {name}:{canister_id} in pulled.json."))?;
+        match &pulled_canister.name {
+            Some(s) if s == name => continue,
+            Some(other_name) => bail!(
+                "{canister_id} is \"{name}\" in dfx.json, but it has name \"{}\" in pulled.json.",
+                other_name
+            ),
+            None => bail!(
+                "{canister_id} is \"{name}\" in dfx.json, but it doesn't have name in pulled.json."
+            ),
+        }
+    }
+
+    for (canister_id, pulled_canister) in &pulled_json.canisters {
+        let pulled_canister_path = get_pulled_wasm_path(*canister_id)?;
+        let bytes = dfx_core::fs::read(&pulled_canister_path)?;
+        let hash_cache = Sha256::digest(bytes);
+        let hash_in_json = hex::decode(&pulled_canister.wasm_hash)?;
+        if hash_cache.as_slice() != hash_in_json {
+            let hash_cache = hex::encode(hash_cache.as_slice());
+            let hash_in_json = &pulled_canister.wasm_hash;
+            bail!(
+                "The pulled wasm of {canister_id} has different hash than in pulled.json:
+    The pulled wasm is at {pulled_canister_path:?}. Its hash is:
+        {hash_cache}
+    The expected hash in pulled.json is:
+        {hash_in_json}"
+            );
+        }
+    }
+
+    Ok(())
+}
+
+fn get_deps_dir(project_root: &Path) -> PathBuf {
+    project_root.join("deps")
+}
+
+pub fn get_candid_path_in_project(project_root: &Path, name: &str) -> PathBuf {
+    get_deps_dir(project_root).join(name).with_extension("did")
```

Comment:
Since the canister id of a pulled canister won't change, but the name might (because it's specified by the developer in dfx.json), do you think it would make sense to store the candid file in `deps/<canister id>.did`?